src/main/kotlin/com/tschuchort/compiletesting/SourceFile.kt

```diff
@@ -30,12 +30,17 @@ abstract class SourceFile {
         /**
          * Compile an existing source file
          */
-        fun fromPath(path: File) = object : SourceFile() {
+        fun fromPath(sourceRoot: File, path: File) = object : SourceFile() {
+
             init {
                 require(path.isFile)
             }
 
-            override fun writeIfNeeded(dir: File): File = path
+            override fun writeIfNeeded(dir: File): File {
+                val file = dir.resolve(path.relativeTo(sourceRoot))
+                path.copyTo(file)
```

Comment:
> If there are two source files with the same name but different path

This is actually one of the reasons why we need to pass in the `sourceRoot` parameter. It allows the caller to disambiguate between files with the same name (`path.relativeTo(sourceRoot)`).

Reply:
Just updated with a test case for this situation.

Reply:
But this way there can still be conflicts when you have two files with the same name but different `sourceRoot`.  So I think it would be better if we changed the way that source files are "discovered" in the two lines I linked above. Or is there some other issue with that solution that I'm not seeing/that necessitates all source files to be in the `sourcesDir`? 
Admittedly, it rarely makes sense to compile two files with the same name and package/location, but it can actually be legal in Kotlin if you change the name of the generated file facade through `@JvmName` (or just declare it to be in a different package than the directory structure would dictate). 

We could simply replace every occurance of `sourcesDir.listFilesRecursively()` with `sources.map { it.path }`.

Reply:
Ok I was thinking that there may have been cases where we'd care to maintain the original directory structure. I'll update with your suggestion.

Reply:
Updated

Reply:
Maybe if it comes to resources. But lets worry about that later, since I didn't consider them before either.